include/fiction/algorithms/physical_design/graph_oriented_layout_design.hpp

```diff
@@ -0,0 +1,1523 @@
+        }
+
+        Ntk::foreach_ci([this](const auto& n) { create_topo_rec(n); });
+    }
+
+  private:
+    void create_topo_rec(const node& n)
+    {
+        /* is permanently marked? */
+        if (this->visited(n) == this->trav_id())
+        {
+            return;
+        }
+
+        bool not_visited = false;
+
+        this->foreach_fanin(n,
+                            [this, &not_visited](const signal& f)
+                            {
+                                if (this->visited(this->get_node(f)) != this->trav_id())
+                                {
+                                    not_visited = true;
+                                };
+                            });
+
+        if (not_visited)
+        {
+            return;
+        }
+
+        /* mark node n permanently */
+        this->set_visited(n, this->trav_id());
+
+        /* visit node */
+        topo_order.push_back(n);
+
+        /* mark children */
+        this->foreach_fanout(n, [this](const auto& fo) { create_topo_rec(fo); });
+    }
+
+    std::vector<node> topo_order;
+};
+/**
+ * Implementation of the graph-oriented layout design algorithm.
+ * This class handles the initialization and execution of the algorithm.
+ *
+ * @tparam Lyt Cartesian gate-level layout type.
+ * @tparam Ntk Network type.
+ */
+template <typename Lyt, typename Ntk>
+class graph_oriented_layout_design_impl
+{
+  public:
+    /**
+     * Constructor for the graph-oriented layout design algorithm.
+     *
+     * @param src The source network to be placed.
+     * @param p The parameters for the graph-enhanced layout search algorithm.
+     * @param st The statistics object to record execution details.
+     */
+    graph_oriented_layout_design_impl(const Ntk& src, const graph_oriented_layout_design_params& p,
+                                      graph_oriented_layout_design_stats& st) :
+            ntk{convert_network<tec_nt>(src)},
+            ps{p},
+            pst{st},
+            timeout{timeout_settings(ps.timeout, ps.high_effort).timeout},
+            start{std::chrono::high_resolution_clock::now()},
+            num_search_space_graphs{ps.high_effort ? 12u : 2u},
+            ssg_vec(num_search_space_graphs)
+    {
+        ntk.substitute_po_signals();
+    }
+    /**
+     * Executes the graph-oriented layout design algorithm and returns the best found layout.
+     *
+     * @return The best layout found by the algorithm.
+     */
+    Lyt run() noexcept
+    {
+        // measure run time
+        mockturtle::stopwatch stop{pst.time_total};
+
+        // Initialize layout to keep track of current best solution
+        Lyt best_lyt{{}, twoddwave_clocking<Lyt>()};
+
+        // Initial setup for networks and nodes to place
+        initialize_networks_and_nodes_to_place();
+
+        // Initial setup for PIs and cost
+        initialize_pis_cost_and_num_expansions();
+
+        // Initial setup for frontiers flags
+        initialize_frontiers_flags();
```

Comment:
I'm confused by the fact that we have to call multiple initialize functions (potentially in the exactly correct order) here. That usually indicates that something is wrong with the API. I know it's only an internal API here. But nevertheless, it is error-prone. Can we "hide" all these functions in a single new `initialize` function? Or just merge them into one as some of them are really short?

But also, maybe we should be thinking about the architecture here a little more first. Calling functions that simply change the status of the underlying object, thus "preparing" the state of the object, is a very C-like thing to do. In C++, you would rather have the initialize function return the ready set-up search space graphs that you then use to initialize your member object with.